microsoft/testsuites/nested/common.py

```diff
@@ -72,28 +76,108 @@ def connect_nested_vm(
     )
 
     # setup connection to nested vm
+    connection_info = ConnectionInfo(
+        address=host.public_address,
+        port=guest_port,
+        username=guest_username,
+        password=guest_password,
+    )
+
     nested_vm = RemoteNode(Node(name=name), 0, name)
     nested_vm.set_connection_info(
-        public_address=host.public_address,
+        **fields_to_dict(connection_info, ["address", "port", "username", "password"])
+    )
+
+    # wait for nested vm ssh connection to be ready
+    try_connect(connection_info)
+
+    return nested_vm
+
+
+def hyperv_connect_nested_vm(
+    host: RemoteNode,
+    guest_username: str,
+    guest_password: str,
+    port: int,
+    guest_image_url: str,
+    name: str = "l2_vm",
+    image_name: str = HYPERV_NESTED_VM_IMAGE_NAME,
+    switch_name: str = "nestedvmswitch",
+    nat_name: str = HYPERV_NAT_NAME,
+) -> RemoteNode:
+    # delete vm if it exists, otherwise it will fail to delete
+    # any present images
+    hyperv = host.tools[HyperV]
+    hyperv.delete_vm(name)
+
+    # Download nested vm image
+    image_name = f"{name}_{image_name}"
+    image_folder_path = "C:\\Users\\lisatest"
+    file_path = host.tools[Wget].get(
+        guest_image_url,
+        image_folder_path,
+        image_name,
+    )
+
+    # setup NAT
+    hyperv.setup_nat_networking(switch_name, nat_name)
+    hyperv.create_vm(
+        name,
+        file_path,
+        switch_name,
+    )
+
+    # cleanup all existing port forwarding rules and
+    # enable port forwarding for the nested vm
+    local_ip = hyperv.get_ip_address(name)
+    hyperv.delete_port_forwarding(nat_name)
+    hyperv.setup_port_forwarding(nat_name, port, local_ip)
+
+    # setup connection to nested vm
+    connection_info = ConnectionInfo(
+        address=host.public_address,
+        port=port,
         username=guest_username,
         password=guest_password,
-        public_port=guest_port,
-        port=guest_port,
     )
 
-    # wait for nested vm ssh connection to be ready
-    try_connect(
-        ConnectionInfo(
-            address=host.public_address,
-            port=guest_port,
-            username=guest_username,
-            password=guest_password,
-        )
+    nested_vm = RemoteNode(Node(name=name), 0, name)
+    nested_vm.set_connection_info(
+        **fields_to_dict(connection_info, ["address", "port", "username", "password"])
     )
+    nested_vm.capability.network_interface = Synthetic()
+
+    # wait for nested vm ssh connection to be ready
+    try_connect(connection_info)
 
     return nested_vm
 
 
+def hyperv_remove_nested_vm(
+    host: RemoteNode,
+    name: str = "L2-VM",
+    image_name: str = HYPERV_NESTED_VM_IMAGE_NAME,
+    switch_name: str = "nestedvmswitch",
+    nat_name: str = "nestedvmnat",
+) -> None:
+    image_name = f"{name}_{image_name}"
+    image_folder_path = "C:\\Users\\lisatest"
```

Comment:
The user name may be changed, it's not reliable. Try to create a folder under system drive.